### PR TITLE
drt: implement access for pins above max routing layers

### PR DESCRIPTION
**Pull request by eder-matheus**

### CODE REVIEW

File: src/drt/test/top_level_term.defok
Comment (maliberty) [inline src/drt/test/top_level_term.defok:37]: Why doesn't the m3 end at the same coordinate as the up vias?

File: src/drt/src/db/obj/frBTerm.h
Comment (maliberty) [inline src/drt/src/db/obj/frBTerm.h:126]: ```bool isAboveTopLayer_ = false;``` avoids the need to repeat it in each ctor.

File: src/drt/src/io/io.cpp
Comment (maliberty) [inline src/drt/src/io/io.cpp:946]: It seems simpler to just have a Rect to fill in rather than building it at the end:

```
Rect bbox = box->getBox();
...
if (...) {
  bbox = viaBox;
}
```

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:968]: Grabbing any via on the right layer may not work well.  At a minimum you should take a default via but even then we may need to add more checks (eg https://github.com/The-OpenROAD-Project/OpenROAD/pull/3056)

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:978]: Is there a reason to front insert?  It will be slower on a vector

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:990]: break?

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:994]: Does this code assume you did pin access before?

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:1000]: If there is an existing route then this is dangerous as you are erasing it.  It seems like an error given the intended use model.

File: src/drt/src/TritonRoute.cpp
Comment (maliberty) [inline src/drt/src/TritonRoute.cpp:1008]: Each via doesn't need to be in a separate path as they are stacked

File: src/drt/src/TritonRoute.tcl
Comment (maliberty) [inline src/drt/src/TritonRoute.tcl:460]: This is mandatory so drop the []

File: src/drt/src/TritonRoute.tcl
Comment (maliberty) [inline src/drt/src/TritonRoute.tcl:474]: Just reduce this to a check and an error if unset

File: src/drt/src/TritonRoute.tcl
Comment (maliberty) [inline src/drt/src/TritonRoute.tcl:481]: Give it a unique id

**Review by maliberty — CHANGES_REQUESTED** [review]

File: src/drt/src/TritonRoute.cpp
Comment (eder-matheus) [reply src/drt/src/TritonRoute.cpp:994]: The `else` part intends to cover the case where we don't have PA information on the bterms.

File: src/drt/test/top_level_term.defok
Comment (eder-matheus) [reply src/drt/test/top_level_term.defok:37]: DRT chooses another position for the met3 via shape, while the vias are placed using the center of the pin shape:
image

File: src/drt/src/TritonRoute.cpp
Comment (eder-matheus) [reply src/drt/src/TritonRoute.cpp:968]: It's not clear how to get the default vias. To get it from DRT, it is necessary to run `initDesign` first, but my new command runs before it is initialized.

We also have this process on grt, but I think this method doesn't fit here:

```
for (odb::dbTechVia* via : vias) {
    odb::dbStringProperty* prop
        = odb::dbStringProperty::find(via, "OR_DEFAULT");

    if (prop == nullptr) {
      continue;
    } else {
      debugPrint(logger_,
                 utl::GRT,
                 "l2v_pitch",
                 1,
                 "Default via: {}.",
                 via->getConstName());
      default_vias[via->getBottomLayer()->getRoutingLevel()] = via;
    }
  }
```

Comment (eder-matheus) [reply src/drt/src/TritonRoute.cpp:968]: Actually, I think it can be simplified. Perhaps we can drop this new command and just run the stack vias process inside drt, without user input. It would be simpler for users and for the implementation, as it can be executed after defining the default vias.

**Comment by eder-matheus** [comment]
@maliberty I think I'm done with the fixes in this PR, and it's ready to merge. I could simplify the code and stack the vias during DRT, before parsing the whole design. I think I've addressed all your comments as well.

File: src/drt/src/TritonRoute.i
Comment (maliberty) [inline src/drt/src/TritonRoute.i:242]: Is this just for testing or is it obsolete?  I don't any tcl caller.

Comment (eder-matheus) [reply src/drt/src/TritonRoute.i]: Obsolete. I just removed it.

File: src/drt/src/TritonRoute.cpp
Comment (osamahammad21) [reply src/drt/src/TritonRoute.cpp:1033]: I might be wrong, so could you please modify the test case to have pin access?

**Review by osamahammad21 — CHANGES_REQUESTED** [review]

File: src/drt/src/io/io.cpp
Comment (eder-matheus) [reply src/drt/src/io/io.cpp:944]: Your comment made me notice that we can have a bug here. Specifically, when we have a net with multiple pins above the max routing layer, this code is not assigning the correct via box to the pins. I believe that this part **must be** dependent on the port itself to work properly. I'm updating it now.

File: src/drt/src/TritonRoute.cpp
Comment (eder-matheus) [reply src/drt/src/TritonRoute.cpp:1033]: You're right; even when running pin_access before detailed_route, this function uses the pin center to place the vias. It's ok to use the center of the pin, since the use case of this new functionality is for the bumps in the upper layers, which usually have a large area. We can drop this part of the code and always use the pin center. @maliberty, do you agree with this approach?

Comment (maliberty) [reply src/drt/src/TritonRoute.cpp]: works for me

File: src/drt/src/io/io.cpp
Comment (osamahammad21) [reply src/drt/src/io/io.cpp]: I don't think this part is handled. you are still duplicating the same pin fig for each port that intersects with the via box. you should add the pinfig of the viabox on the top routing layer only once.

Comment (eder-matheus) [reply src/drt/src/io/io.cpp]: I think now it's handled. The code only iterates over all the bpins when the layer is below the max routing layer.

**Comment by eder-matheus** [comment]
@maliberty @osamahammad21 Is there anything else I need to change in this PR?

**Comment by maliberty** [comment]
@osamahammad21 please give a final review and approve if satisfied.  I have no further comments

**Review by osamahammad21 — APPROVED** [review]
@eder-matheus Could you please check the ci failure?

**Comment by eder-matheus** [comment]
> @eder-matheus Could you please check the ci failure?

Not sure what caused the failure, but all the tests are passing. I think the pipeline for this run was killed before it completes.

**Comment by eder-matheus** [comment]
@maliberty I think I need your approval too because you have requested changes.

**Review by maliberty — APPROVED** [review]

**Pull request merged**
